### PR TITLE
One pass autodetect encodings

### LINKED ISSUE #124: Build fast one-pass "detect encoding function"
Currently we might do something like this...

```C++
simdutf_warn_unused encoding_type implementation::autodetect_encoding(const char * input, size_t length) const noexcept {
    // If there is a BOM, then we trust it.
    auto bom_encoding = simdutf::BOM::check_bom(input, length);
    if(bom_encoding != encoding_type::unspecified) { return bom_encoding; }
    // UTF8 is common, it includes ASCII, and is commonly represented
    // without a BOM, so if it fits, go with that. Note that it is still
    // possible to get it wrong, we are only 'guessing'. If some has UTF-16
    // data without a BOM, it could pass as UTF-8.
    //
    // An interesting twist might be to check for UTF-16 ASCII first (every
    // other byte is zero).
    if(validate_utf8(input, length)) { return encoding_type::UTF8; }
    // The next most common encoding that might appear without BOM is probably
    // UTF-16LE, so try that next.
    if((length % 2) == 0) {
      if(validate_utf16(reinterpret_cast<const char16_t*>(input), length/2)) { return encoding_type::UTF16_LE; }
    }
    if((length % 4) == 0) {
      if(validate_utf32(reinterpret_cast<const char32_t*>(input), length/4)) { return encoding_type::UTF32_LE; }
    }
    return encoding_type::unspecified;
}
```

It would not normally be so bad, but our SIMD `validate_utf*` functions are optimistic: they expect the input to be valid and so they always scan the whole input. If the input is UTF-8, it is pretty good...  We could use our fallback functions instead... 

Importantly, the 'answer' is not well defined. What we would like is a function that returns which unicode types are *possible* given the input. So the result is an enumeration... utf8, utf16_le, utf16_be, utf32_le, utf32_be....?

But it seems that there ought to be a one-pass algorithm to detect the unicode type... So you scan the input, using a single pass, and when you are done, you know exactly which types are allowed.

### PR DESCRIPTION
**Pull request by NicolasJiaxin**
Fix #124.

### CODE REVIEW

**Comment by lemire** [comment]
Very exciting !!!

**Comment by NicolasJiaxin** [comment]
The current algorithm is not entirely correct. Even if the surrogate_bitmask is not zero, the input could still be UTF-32LE depending on the positions of the surrogate.

**Comment by NicolasJiaxin** [comment]
By surrogate, I mean that the high-byte of a 16-bit word is 0xd8XX. Basically, in the presence of surrogate, it can only be UTF-16LE **or** UTF-32LE. To be valid UTF16-LE, each surrogate must be preceded/followed by another surrogate. If that is the case, then it cannot be UTF32-LE since one of those will form the upper 2 bytes of a 32-bit word and this word would be too big. On the other hand, even if we detect a surrogate, it still could be UTF-32LE. For example, the 32-bit word 0x0001d800 is valid UTF-32LE, but the surrogate_bitmask would not be zero. However, one thing is sure, the input cannot be UTF-8 in the presence of surrogates.

**Comment by lemire** [comment]
I think that returning `std::vector`, though fine, might be an overkill.

Suppose we do the following...

```
enum encoding_type {
        UTF16_LE  = 1,   // BOM 0xff 0xfe
        UTF16_BE  = 2,   // BOM 0xfe 0xff
        UTF32_LE  = 4,   // BOM 0xff 0xfe 0x00 0x00
        UTF32_BE  = 8,   // BOM 0x00 0x00 0xfe 0xff
        UTF8  = 16,       // BOM 0xef 0xbb 0xbf
        unspecified  = 0 
};
```

Then you could just return an `int`. Instead of 

```
{ out.push_back(encoding_type::UTF32_LE); }
```

We could just do

```
out |= encoding_type::UTF32_LE;
```


To check that it might be UTF8, someone could do...

```C++
auto type = detect_encodings(...);
if (type | encoding_type::UTF8) {
  // could be UTF8
}
```

(I would simplify your function name from `op_autodetect_encodings`  to something simplier.)

In practice, the `std::vector` return type is probably not going to carry a measurable overhead... but it makes the code subject to heap-based memory allocations which is a real shame! My proposal would be (maybe not measurably more) efficient and allocation-free.

**Comment by NicolasJiaxin** [comment]
I agree std::vector is probably a bit much and your proposal to return an int more elegant. Just to be sure though, instead of this,

> ```c++
> if (type | encoding_type::UTF8) {
> ```

I think you meant 

```c++
if (type & encoding_type::UTF8) {
 ```
right? (To check if the UTF-8 bit is turned on).

**Comment by lemire** [comment]
Of course, my version would always return true.

**Comment by NicolasJiaxin** [comment]
@lemire I have added a few valid cases that cover most scenarios that are interesting, I think. Let me know if you have others.

For now, this is my algorithm:
In each iteration, you check if there is any surrogate in the high byte of each 16-bit word. If there is not, then validate UTF-8 and UTF-32 as usual. If there is, then the input cannot be UTF-8 for sure. From that point, it can only be either UTF-16 or UTF-32 as I explain above. To make that distinction, we check if there are any surrogates in the 16-bit upper word of each 32-bit word. If there is, then it can only be UTF-16 and we continue to process the input just validating UTF-16. If there is not, then it can only be UTF-32 and we finish processing the data just validating UTF-32. My code is a little bit ugly, but this is the idea.

I will work on benchmarking now.

**Comment by NicolasJiaxin** [comment]
Here are the numbers I get with haswell on a few files, 
||validate_utf*|detect_encodings|
|-|-|-|
|Arabic-Lipsum.utf8.txt|12.146|12.165|
|Chinese-Lipsum.utf8.txt|12.169|9.950|
|Emoji-Lipsum.utf8.txt|11.878|9.967|
|Latin-Lipsum.utf8.txt|89.629|25.069|
|Russian-Lipsum.utf8.txt|12.098|9.866|
|Arabic-Lipsum.utf16.txt|17.793|24.805|
|Chinese-Lipsum.utf16.txt|19.966|11.048|
|Emoji-Lipsum.utf16.txt|8.862|0.619|
|Latin-Lipsum.utf16.txt|17.094|24.350|
|Russian-Lipsum.utf16.txt|16.651|24.589|
|Arabic-Lipsum.utf32.txt|8.573|24.764|
|Chinese-Lipsum.utf32.txt|9.143|9.882|
|Emoji-Lipsum.utf32.txt|9.013|9.938|
|Latin-Lipsum.utf32.txt|8.377|23.593|
|Russian-Lipsum.utf32.txt|8.954|23.811|

On UTF-8, we get a decrease which is expected. However, on UTF-16 and UTF-32, we get better results on detect_encodings most of the time. Then, there is Emoji-Lipsum.utf16.txt. I will investigate, there is most certainly something wrong there.

**Comment by lemire** [comment]
These are good numbers (except for `Emoji-Lipsum.utf16.txt`).

**Comment by NicolasJiaxin** [comment]
On a ROME processor, I get,
```
nicolas@programming_station-for-nicolas:~/simdutf$ ./build/benchmarks/benchmark -P detect_encodings -F ~/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt 
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 66.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
detect_encodings+fallback, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   4.066 GB/s (0.9 %)    4.066 Gc/s     1.00 byte/char 
detect_encodings+haswell, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   2.982 GB/s (0.2 %)    2.982 Gc/s     1.00 byte/char 
detect_encodings+westmere, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   4.391 GB/s (0.3 %)    4.391 Gc/s     1.00 byte/char 
nicolas@programming_station-for-nicolas:~/simdutf$ ./build/benchmarks/benchmark -P validate_utf16 -F ~/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt 
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 66.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
validate_utf16+fallback, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   1.694 GB/s (0.3 %)    0.847 Gc/s     2.00 byte/char 
validate_utf16+haswell, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
  10.727 GB/s (0.7 %)    5.364 Gc/s     2.00 byte/char 
validate_utf16+westmere, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   8.793 GB/s (1.1 %)    4.397 Gc/s     2.00 byte/char 
```
which is not great, but better already. I have not found the issue. The code is executing as expected with Emoji-Lipsum.utf16.txt.

File: src/implementation.cpp [hunk 1]
```diff
@@ -89,6 +89,10 @@ class detect_best_supported_implementation_on_first_use final : public implement
   const std::string &description() const noexcept final { return set_best()->description(); }
   uint32_t required_instruction_sets() const noexcept final { return set_best()->required_instruction_sets(); }
 
+  simdutf_warn_unused int detect_encodings(const char * input, size_t length) const noexcept {
```
Comment (lemire) [inline src/implementation.cpp:92]:
```suggestion
  simdutf_warn_unused int detect_encodings(const char * input, size_t length) const noexcept override {
```

File: src/implementation.cpp [hunk 2]
```diff
@@ -214,6 +218,10 @@ const std::initializer_list<const implementation *> available_implementation_poi
 // So we can return UNSUPPORTED_ARCHITECTURE from the parser when there is no support
 class unsupported_implementation final : public implementation {
 public:
+  simdutf_warn_unused int detect_encodings(const char *, size_t) const noexcept {
```
Comment (lemire) [inline src/implementation.cpp:221]:
```suggestion
  simdutf_warn_unused int detect_encodings(const char *, size_t) const noexcept override {
```

File: src/arm64/arm_detect_encodings.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                        const auto sh = simd8<uint8_t>({1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,0xFF});
+                        const auto vHshifteddown = vH.apply_lookup_16_to(sh);
+                        const auto match = vHshifteddown == vL;
+
+                        const auto fmatch = simd8<bool>(simd8<uint8_t>(match) | sh);
+
+                        if (fmatch.all() && low_vh == 0) {
+                            input += (high_vl == 0) ? 16 : 15;
+                        } else {
+                            printf("A");
```
Comment (lemire) [inline src/arm64/arm_detect_encodings.cpp:107]:
```suggestion
```

**Comment by lemire** [comment]
On the UTF-16 emoji file, I am getting 'poor' results on AMD Rome:

```
$ ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P detect
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 67.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
detect_encodings+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   4.754 ins/byte,    4.754 cycle/byte,    3.386 GB/s (24.1 %),     3.403 GHz,    4.731 ins/cycle
   4.754 ins/char,    4.754 cycle/char,    3.386 Gc/s (24.1 %)     1.00 byte/char
WARNING: Measurements are noisy, try increasing iteration count (-I).
detect_encodings+haswell, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   5.551 ins/byte,    5.551 cycle/byte,    2.991 GB/s (0.7 %),     3.403 GHz,    4.880 ins/cycle
   5.551 ins/char,    5.551 cycle/char,    2.991 Gc/s (0.7 %)     1.00 byte/char
detect_encodings+westmere, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   5.691 ins/byte,    5.691 cycle/byte,    3.258 GB/s (0.5 %),     3.403 GHz,    5.449 ins/cycle
   5.691 ins/char,    5.691 cycle/char,    3.258 Gc/s (0.5 %)     1.00 byte/char
```

Also on ARM:

```
❯  ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P detect        
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 69.1
testcases: 1
input detected as UTF16 little-endian
current system detected as arm64
===========================
detect_encodings+arm64, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
kpc_set_config failed, run the program with sudo
   2.194 GB/s (1.9 %)    2.194 Gc/s     1.00 byte/char 
detect_encodings+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   5.141 GB/s (4.6 %)    5.141 Gc/s     1.00 byte/char 
```

Regarding ARM, I created a new issue: https://github.com/simdutf/simdutf/issues/136 .

The expensive component is something you borrowed (rightly so) from our 'validate UTF16' routine:

```C++
                while (input + simd16<uint16_t>::ELEMENTS * 2 < end16) {
                    const auto in0 = simd16<uint16_t>(input);
                    const auto in1 = simd16<uint16_t>(input + simd16<uint16_t>::ELEMENTS);
                    const auto t0 = in0.shr<8>();
                    const auto t1 = in1.shr<8>();
                    const auto in_16 = simd16<uint16_t>::pack(t0, t1);
                    const auto surrogates_wordmask = (in_16 & v_f8) == v_d8;
                    const uint32_t surrogates_bitmask = surrogates_wordmask.to_bitmask();
                    if (surrogates_bitmask == 0x0) {
                        input += simd16<uint16_t>::ELEMENTS * 2;
                    } else {
                        const uint32_t V = ~surrogates_bitmask;
                        const auto    vH = (in_16 & v_fc) == v_dc;
                        const uint32_t H = vH.to_bitmask();
                        const uint32_t L = ~H & surrogates_bitmask;
                        const uint32_t a = L & (H >> 1);
                        const uint32_t b = a << 1;
                        const uint32_t c = V | a | b;
                        if (c == 0xffffffff) {
                            input += simd16<uint16_t>::ELEMENTS * 2;
                        } else if (c == 0x7fffffff) {
                            input += simd16<uint16_t>::ELEMENTS * 2 - 1;
                        } else {
                            return simdutf::encoding_type::unspecified;
                        }
                    }
```

**Comment by lemire** [comment]
I am puzzled regarding the speed of the fallback routine.  Let us look at your numbers...

```
detect_encodings+fallback, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   4.066 GB/s (0.9 %)    4.066 Gc/s     1.00 byte/char 

validate_utf16+fallback, input size: 65542, iterations: 2000, dataset: /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   1.694 GB/s (0.3 %)    0.847 Gc/s     2.00 byte/char 
```


How is that possible?

**Comment by lemire** [comment]
I am also getting the same result. It does not sound possible... ?

```
$ ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P "detect_encodings+fallback" -P "validate_utf16+fallback"
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 67.1
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
detect_encodings+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   4.754 ins/byte,    4.754 cycle/byte,    3.386 GB/s (0.2 %),     3.402 GHz,    4.731 ins/cycle
   4.754 ins/char,    4.754 cycle/char,    3.386 Gc/s (0.2 %)     1.00 byte/char
validate_utf16+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
   9.507 ins/byte,    9.507 cycle/byte,    2.026 GB/s (4.3 %),     3.404 GHz,    5.660 ins/cycle
  19.012 ins/char,   19.012 cycle/char,    1.013 Gc/s (4.3 %)     2.00 byte/char
```

**Comment by lemire** [comment]
If the file is valid UTF16, then there is no way that validate_utf16 can be slower than detect_encodings. Can it?

**Comment by lemire** [comment]
@NicolasJiaxin 

To help understand the problem, we should make `run_detect_encodings` smarter. In particular, we should try to detect a BOM and eat it up if it is present. Also, we should print the result so we can debug.

Try the following patch or the equivalent...


```diff
index 71e5ecf..0859762 100644
--- a/benchmarks/src/benchmark.cpp
+++ b/benchmarks/src/benchmark.cpp
@@ -1391,20 +1391,52 @@ void Benchmark::run_count_utf16(const simdutf::implementation& implementation, s
 }
 
 void Benchmark::run_detect_encodings(const simdutf::implementation& implementation, size_t iterations) {
-    const char*  data = reinterpret_cast<const char*>(input_data.data());
-    const size_t size = input_data.size();
+    const simdutf::encoding_type bom  = BOM::check_bom(input_data.data(), input_data.size());
+    const char* data = reinterpret_cast<const char*>(input_data.data() + BOM::bom_byte_size(bom));
+    const size_t size = input_data.size() - BOM::bom_byte_size(bom);
     volatile size_t sink{0};
-
     auto proc = [&implementation, data, size, &sink]() {
         sink = implementation.detect_encodings(data, size);
     };
     count_events(proc, iterations); // warming up!
     const auto result = count_events(proc, iterations);
-    if((sink == 0) && (size != 0) && (iterations > 0)) { std::cerr << "The output is zero which might indicate an error.\n"; }
     size_t char_count = size;
+    if((sink == 0) && (size != 0) && (iterations > 0)) {
+        std::cerr << "The output is zero which might indicate an error.\n";
+    } else {
+        std::cout << "Detected format: ";
+        if(sink & simdutf::encoding_type::UTF8) {
+            char_count =  active_implementation->count_utf8(data, size);
+            std::cout << " UTF8";
+        }
+        if(sink & simdutf::encoding_type::UTF16_LE) {
+            std::cout << " UTF16LE";
+            char_count =  active_implementation->count_utf16(reinterpret_cast<const char16_t*>(data), size/2);
+        }
+        if(sink & simdutf::encoding_type::UTF32_LE) {
+            std::cout << " UTF32LE";
+            char_count = size/4;
+        }
+        std::cout << std::endl;
+    }
+    if((bom) && (bom & ~sink)) {
+        std::cerr << "[Error] BOM format     : ";
+        if(bom & simdutf::encoding_type::UTF8) {
+            std::cerr << " UTF8";
+        } else if(bom & simdutf::encoding_type::UTF16_LE) {
+            std::cerr << " UTF16LE";
+        } else if(bom & simdutf::encoding_type::UTF32_LE) {
+            std::cerr << " UTF32LE";
+        }
+        std::cerr << std::endl;
+    }
+    if((sink & (sink - 1)) != 0) {
+        std::cout << "More than one format possible, character count is ambiguous." << std::endl;
+    }
     print_summary(result, size, char_count);
 }
 
 const std::set<std::string>& Benchmark::all_procedures() const {
     return known_procedures;
 }

```

**Comment by lemire** [comment]
@NicolasJiaxin What are your results now... do we have more meaningful results?

**Comment by lemire** [comment]
It is a fun problem. :-)

**Comment by NicolasJiaxin** [comment]
The problem was not in the main loop. It was because sometimes the loop exited in the middle of a UTF-8 codepoint and instead of validating with the utf8_checker, I was using the scalar validation for the tail which started at a continuation byte. I added a test to check for this.

**Comment by lemire** [comment]
> numbers are still down on Emoji-Lipsum.utf16.txt.



Here are the numbers I am getting...

```
 ./build/benchmarks/benchmark -F unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt -P detect
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
testcases: 1
input detected as UTF16 little-endian
current system detected as haswell
===========================
detect_encodings+fallback, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
Detected format:  UTF16LE
   2.503 GB/s (0.7 %)    0.626 Gc/s     4.00 byte/char 
detect_encodings+haswell, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
Detected format:  UTF16LE
   3.176 GB/s (2.9 %)    0.794 Gc/s     4.00 byte/char 
detect_encodings+westmere, input size: 65542, iterations: 2000, dataset: unicode_lipsum/lipsum/Emoji-Lipsum.utf16.txt
Detected format:  UTF16LE
   2.922 GB/s (3.5 %)    0.731 Gc/s     4.00 byte/char 
```

It is, admittedly, way down compared to other files, but at least the performance numbers appear credible... don't they?

**Comment by NicolasJiaxin** [comment]
> It is, admittedly, way down compared to other files, but at least the performance numbers appear credible... don't they?

It looks credible. Where are you getting these numbers?

**Comment by lemire** [comment]
@NicolasJiaxin I just grabbed your code and ran it on my mac (clang 12).

**Comment by NicolasJiaxin** [comment]
Ah with clang, I see. Yes this looks better and credible.

**Comment by lemire** [comment]
@NicolasJiaxin What are your latest numbers?

File: src/arm64/arm_detect_encodings.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        const auto v3 = u3.shr<8>();
+
+        const auto in16 = simd16<uint16_t>::pack(v0, v1);
+        const auto nextin16 = simd16<uint16_t>::pack(v2, v3);
+
+        const auto surrogates_wordmask0 = (in16 & v_f8) == v_d8;
+        const auto surrogates_wordmask1 = (nextin16 & v_f8) == v_d8;
+
+        // Check for surrogates
+        if (!surrogates_wordmask0.none() || !surrogates_wordmask1.none()) {
```
Comment (lemire) [inline src/arm64/arm_detect_encodings.cpp:43]: Under ARM, the 'none()' method is very slow. Please consider switching over to @[danlark1](https://github.com/danlark1)'s  to_bitmask64() which is at least twice as fast... See https://github.com/simdutf/simdutf/pull/145 for my validate_utf16 benchmarks where it made a huge difference.

**Comment by NicolasJiaxin** [comment]
On ARM, comparison of old validation with new validation. We see improvements all around.


||old (GB/s)|new (GB/s)|
|-|-|-|
|Arabic-Lipsum.utf8.txt|6.602|6.608|
|Chinese-Lipsum.utf8.txt|5.097|5.405|
|Emoji-Lipsum.utf8.txt|5.137|5.375|
|Latin-Lipsum.utf8.txt|11.932|13.822|
|Russian-Lipsum.utf8.txt|5.122|5.421|
|Arabic-Lipsum.utf16.txt|11.721|13.679|
|Chinese-Lipsum.utf16.txt|5.093|5.315|
|Emoji-Lipsum.utf16.txt|1.406|2.118|
|Latin-Lipsum.utf16.txt|11.766|13.735|
|Russian-Lipsum.utf16.txt|11.744|13.713|
|Arabic-Lipsum.utf32.txt|11.956|13.907|
|Chinese-Lipsum.utf32.txt|5.108|5.421|
|Emoji-Lipsum.utf32.txt|5.095|5.369|
|Latin-Lipsum.utf32.txt|12.147|13.913|
|Russian-Lipsum.utf32.txt|12.480|13.905|

**Pull request merged**